File: farmopti/first_demo/lib/optimizer/decisionAnalysis/extractDecisions.ts
import { CANDIDATE_CONFIG, CONFIG } from "../config";
import { loadExternalVariables } from "../externalVariables";
import { FieldSimulator } from "../fieldSimulator";
import { ceilToHour, combineDateAndTime, dateOnlyOf, floorToHour, formatIso } from "../datetime";
import { EXTERNAL_DIR } from "../paths";
import type {
  Candidate,
  ExternalVariables,
  FieldOption,
  ManagementPlanRow,
  OptimizationSummary,
  ScheduleRow,
} from "../types";

export interface StateChange {
  before: unknown;
  after: unknown;
  change?: number;
}

export interface StateTransitionRecord {
  plan_id: string;
  candidate_id: string;
  field_id: string;
  operation: string;
  target: string;
  start_time: string;
  end_time: string;
  state_before: Record<string, unknown>;
  state_after: Record<string, unknown>;
  state_changes: Record<string, StateChange>;
  direct_revenue_aud: number;
  direct_cost_aud: number;
  direct_cash_effect_aud: number;
  state_yield_effect_t_ha: number;
}

export interface AlternativeTime {
  candidate_id: string;
  start_time: string;
  direct_cash_effect_aud: number;
}

export interface TimingInfo {
  feasible_candidate_count: number;
  earliest_feasible_time: string | null;
  latest_feasible_time: string | null;
  selected_time: string | null;
  selected_candidate_id: string | null;
  alternative_times: AlternativeTime[];
}

export interface ActionRecord {
  decision_id: string;
  type: "ACTION_SELECTED" | "ACTION_SKIPPED";
  plan_id: string;
  field_id: string;
  operation: string;
  target: string;
  required: boolean;
  selected: boolean;
  depends_on: string;
  min_gap_hours: number;
  timing: TimingInfo;
  financial: {
    direct_revenue_aud: number;
    direct_cost_aud: number;
    direct_cash_effect_aud: number;
    state_yield_effect_t_ha: number;
  } | null;
  state_transition: StateTransitionRecord | null;
  counterfactual: any;
  importance: number;
}

export interface FieldDecisionRecord {
  decision_id: string;
  type: "FIELD_OPTION_SELECTED";
  field_id: string;
  selected_option_id: string;
  selected_objective_value_aud: number;
  selected_direct_cash_effect_aud: number;
  selected_terminal_value_aud: number;
  best_local_alternative: {
    option_id: string;
    objective_value_aud: number;
    local_value_difference_aud: number;
  } | null;
}

export interface ResourceRecord {
  decision_id: string;
  type: "RESOURCE_UTILISATION";
  resource_type: "machine" | "water" | "labour";
  resource_id?: string;
  date?: string;
  time?: string;
  used_hours?: number;
  available_hours?: number;
  used_ml?: number;
  capacity_ml?: number;
  used_workers?: number;
  available_workers?: number;
  utilisation: number;
  binding: boolean;
  affected_plans?: string[];
}

export interface DecisionTrace {
  generated_at: string;
  summary: OptimizationSummary;
  actions: ActionRecord[];
  field_decisions: FieldDecisionRecord[];
  resources: ResourceRecord[];
  counterfactuals: any[];
}

function stateDiff(before: Record<string, unknown>, after: Record<string, unknown>): Record<string, StateChange> {
  const changes: Record<string, StateChange> = {};
  const keys = new Set([...Object.keys(before), ...Object.keys(after)]);

  for (const key of [...keys].sort()) {
    const a = before[key];
    const b = after[key];

    if (typeof a === "number" && typeof b === "number") {
      const delta = Math.round((b - a) * 1e4) / 1e4;
      if (Math.abs(delta) > 1e-9) {
        changes[key] = { before: Math.round(a * 1e4) / 1e4, after: Math.round(b * 1e4) / 1e4, change: delta };
      }
    } else if (a !== b) {
      changes[key] = { before: a, after: b };
    }
  }

  return changes;
}

function simulateSelectedOption(data: ExternalVariables, option: FieldOption): StateTransitionRecord[] {
  const fieldId = String(option.field_id);
  const planMap = new Map(data.management.map((p) => [p.plan_id, p]));
  const simulator = new FieldSimulator(data, CONFIG, fieldId);
  const records: StateTransitionRecord[] = [];

  const actions = [...option.actions].sort((a, b) => a.start_time - b.start_time);

  for (const action of actions) {
    const planId = String(action.plan_id);
    const plan = planMap.get(planId)!;

    simulator.advanceTo(action.start_time);
    const before = simulator.snapshot() as unknown as Record<string, unknown>;
    const result = simulator.evaluateAndApply(action, plan);
    if (result === null) continue;

    const after = result.state_after as unknown as Record<string, unknown>;

    records.push({
      plan_id: planId,
      candidate_id: String(action.candidate_id),
      field_id: fieldId,
      operation: String(action.operation),
      target: action.target ?? "",
      start_time: formatIso(action.start_time),
      end_time: formatIso(action.end_time),
      state_before: before,
      state_after: after,
      state_changes: stateDiff(before, after),
      direct_revenue_aud: result.direct_revenue_aud,
      direct_cost_aud: result.direct_cost_aud,
      direct_cash_effect_aud: result.direct_cash_effect_aud,
      state_yield_effect_t_ha: result.state_yield_effect_t_ha,
    });
  }

  return records;
}

export function buildActionRecords(
  data: ExternalVariables,
  candidates: Candidate[],
  options: FieldOption[],
  schedule: ScheduleRow[],
  summary: OptimizationSummary,
): ActionRecord[] {
  const selectedOptionIds = new Set(summary.selected_options.map((o) => String(o.option_id)));
  const selectedOptions = options.filter((o) => selectedOptionIds.has(String(o.option_id)));

  const transitions = new Map<string, StateTransitionRecord>();
  for (const option of selectedOptions) {
    for (const record of simulateSelectedOption(data, option)) {
      transitions.set(record.plan_id, record);
    }
  }

  const candidatesByPlan = new Map<string, Candidate[]>();
  for (const c of candidates) {
    const list = candidatesByPlan.get(c.plan_id) ?? [];
    list.push(c);
    candidatesByPlan.set(c.plan_id, list);
  }

  const records: ActionRecord[] = [];

  for (const plan of data.management) {
    const planId = plan.plan_id;
    const selectedRows = schedule.filter((s) => s.plan_id === planId);
    const selected = selectedRows.length > 0;
    const selectedRow = selected ? selectedRows[0] : null;

    const planCandidates = [...(candidatesByPlan.get(planId) ?? [])].sort((a, b) => a.start_time - b.start_time);

    const timing: TimingInfo = {
      feasible_candidate_count: planCandidates.length,
      earliest_feasible_time: null,
      latest_feasible_time: null,
      selected_time: null,
      selected_candidate_id: null,
      alternative_times: [],
    };

    if (planCandidates.length > 0) {
      timing.earliest_feasible_time = formatIso(planCandidates[0].start_time);
      timing.latest_feasible_time = formatIso(planCandidates[planCandidates.length - 1].start_time);
    }

    if (selected && selectedRow) {
      timing.selected_time = formatIso(selectedRow.start_time);
      timing.selected_candidate_id = String(selectedRow.candidate_id);

      const alternatives = planCandidates
        .filter((c) => String(c.candidate_id) !== String(selectedRow.candidate_id))
        .map((c) => ({ candidate: c, distance: Math.abs(c.start_time - selectedRow.start_time) }))
        .sort((a, b) => a.distance - b.distance)
        .slice(0, 5);

      timing.alternative_times = alternatives.map(({ candidate }) => ({
        candidate_id: String(candidate.candidate_id),
        start_time: formatIso(candidate.start_time),
        direct_cash_effect_aud: candidate.direct_cash_effect_aud,
      }));
    }

    const record: ActionRecord = {
      decision_id: `${planId}:action`,
      type: selected ? "ACTION_SELECTED" : "ACTION_SKIPPED",
      plan_id: planId,
      field_id: String(plan.field_id),
      operation: String(plan.operation),
      target: plan.target ?? "",
      required: plan.required,
      selected,
      depends_on: plan.depends_on ?? "",
      min_gap_hours: plan.min_gap_hours ?? 0.0,
      timing,
      financial: null,
      state_transition: transitions.get(planId) ?? null,
      counterfactual: null,
      importance: 0.0,
    };

    if (selected && selectedRow) {
      record.financial = {
        direct_revenue_aud: selectedRow.direct_revenue_aud,
        direct_cost_aud: selectedRow.direct_cost_aud,
        direct_cash_effect_aud: selectedRow.direct_cash_effect_aud,
        state_yield_effect_t_ha: selectedRow.state_yield_effect_t_ha,
      };
    }

    records.push(record);
  }

  return records;
}

export function buildFieldRecords(options: FieldOption[], summary: OptimizationSummary): FieldDecisionRecord[] {
  const selectedIds = new Map(summary.selected_options.map((o) => [String(o.option_id), o]));

  const optionsByField = new Map<string, FieldOption[]>();
  for (const option of options) {
    const list = optionsByField.get(String(option.field_id)) ?? [];
    list.push(option);
    optionsByField.set(String(option.field_id), list);
  }

  const records: FieldDecisionRecord[] = [];

  for (const [fieldId, fieldOptions] of optionsByField) {
    const selected = fieldOptions.find((o) => selectedIds.has(String(o.option_id)))!;

    const alternatives = fieldOptions
      .filter((o) => String(o.option_id) !== String(selected.option_id))
      .sort((a, b) => b.objective_value_aud - a.objective_value_aud);

    const bestAlt = alternatives[0] ?? null;

    records.push({
      decision_id: `${fieldId}:field_option`,
      type: "FIELD_OPTION_SELECTED",
      field_id: fieldId,
      selected_option_id: String(selected.option_id),
      selected_objective_value_aud: selected.objective_value_aud,
      selected_direct_cash_effect_aud: selected.total_direct_cash_effect_aud,
      selected_terminal_value_aud: selected.terminal_value_aud,
      best_local_alternative: bestAlt
        ? {
            option_id: String(bestAlt.option_id),
            objective_value_aud: bestAlt.objective_value_aud,
            local_value_difference_aud: Math.round((selected.objective_value_aud - bestAlt.objective_value_aud) * 100) / 100,
          }
        : null,
    });
  }

  return records;
}

export function buildResourceRecords(data: ExternalVariables, schedule: ScheduleRow[]): ResourceRecord[] {
  if (schedule.length === 0) return [];

  const records: ResourceRecord[] = [];

  const byMachine = new Map<string, ScheduleRow[]>();
  for (const row of schedule) {
    if (!row.machine_id) continue;
    const list = byMachine.get(row.machine_id) ?? [];
    list.push(row);
    byMachine.set(row.machine_id, list);
  }

  for (const [machineId, group] of byMachine) {
    const usedHours = group.reduce((sum, r) => sum + (r.end_time - r.start_time) / 3_600_000, 0);

    const availabilityRows = data.machineAvailability.filter((a) => a.machine_id === machineId && a.available === 1);
    let availableHours = 0.0;
    for (const row of availabilityRows) {
      if (row.available_from === null || row.available_to === null) continue;
      const start = combineDateAndTime(row.date, row.available_from);
      const end = combineDateAndTime(row.date, row.available_to);
      availableHours += Math.max(0.0, (end - start) / 3_600_000);
    }

    const utilisation = availableHours > 0 ? usedHours / availableHours : 0.0;

    records.push({
      decision_id: `machine:${machineId}`,
      type: "RESOURCE_UTILISATION",
      resource_type: "machine",
      resource_id: String(machineId),
      used_hours: Math.round(usedHours * 100) / 100,
      available_hours: Math.round(availableHours * 100) / 100,
      utilisation: Math.round(utilisation * 1e4) / 1e4,
      binding: utilisation >= 0.9,
      affected_plans: group.map((r) => String(r.plan_id)),
    });
  }

  for (const waterRow of data.water) {
    const used = schedule
      .filter((s) => dateOnlyOf(s.start_time) === waterRow.date)
      .reduce((sum, s) => sum + s.water_ml, 0);

    const capacity = Math.min(waterRow.available_water_ml, waterRow.max_delivery_ml_per_day);
    const utilisation = capacity > 0 ? used / capacity : 0.0;

    records.push({
      decision_id: `water:${waterRow.date}`,
      type: "RESOURCE_UTILISATION",
      resource_type: "water",
      date: waterRow.date,
      used_ml: Math.round(used * 1000) / 1000,
      capacity_ml: Math.round(capacity * 100) / 100,
      utilisation: Math.round(utilisation * 1e4) / 1e4,
      binding: utilisation >= 0.9,
    });
  }

  const timeStepMs = Number(CANDIDATE_CONFIG.time_step_hours) * 3_600_000;
  const start = floorToHour(Math.min(...schedule.map((s) => s.start_time)));
  const end = ceilToHour(Math.max(...schedule.map((s) => s.end_time)));

  for (let t = start; t < end; t += timeStepMs) {
    const date = dateOnlyOf(t);
    const labourRow = data.labour.find((l) => l.date === date);
    if (!labourRow) continue;

    const capacity = labourRow.available_workers;
    const active = schedule.filter((s) => s.start_time <= t && s.end_time > t);
    const used = active.reduce((sum, s) => sum + s.workers_required, 0);
    const utilisation = capacity > 0 ? used / capacity : 0.0;

    if (used > 0 || utilisation >= 0.9) {
      records.push({
        decision_id: `labour:${formatIso(t)}`,
        type: "RESOURCE_UTILISATION",
        resource_type: "labour",
        time: formatIso(t),
        used_workers: used,
        available_workers: capacity,
        utilisation: Math.round(utilisation * 1e4) / 1e4,
        binding: utilisation >= 0.9,
        affected_plans: active.map((s) => String(s.plan_id)),
      });
    }
  }

  return records;
}

export function updateImportance(trace: DecisionTrace): DecisionTrace {
  const baseline = Math.abs(trace.summary.total_objective_value_aud) || 1.0;
  const counterfactualByPlan = new Map(trace.counterfactuals.map((x: any) => [String(x.plan_id), x]));

  for (const record of trace.actions) {
    let score = 0.0;

    if (record.required) score += 0.15;

    const financial = record.financial;
    if (financial && financial.direct_cash_effect_aud < 0) {
      score += Math.min(0.2, (Math.abs(financial.direct_cash_effect_aud) / baseline) * 10);
    }

    const counterfactual = counterfactualByPlan.get(record.plan_id);
    if (counterfactual) {
      record.counterfactual = counterfactual;

      const objectiveChange = counterfactual.objective_change_aud;
      if (objectiveChange !== null && objectiveChange !== undefined) {
        const delta = Math.abs(objectiveChange);
        score += Math.min(0.6, (delta / baseline) * 20);
      } else if (counterfactual.feasible === false) {
        score += 0.6;
      }
    }

    if (!record.required) score += 0.1;

    record.importance = Math.min(1.0, Math.round(score * 1e4) / 1e4);
  }

  return trace;
}

export function generateDecisionTrace(
  candidates: Candidate[],
  options: FieldOption[],
  schedule: ScheduleRow[],
  summary: OptimizationSummary,
  externalVariablesDir: string = EXTERNAL_DIR,
): DecisionTrace {
  const data = loadExternalVariables(externalVariablesDir);

  const trace: DecisionTrace = {
    generated_at: new Date().toISOString(),
    summary,
    actions: buildActionRecords(data, candidates, options, schedule, summary),
    field_decisions: buildFieldRecords(options, summary),
    resources: buildResourceRecords(data, schedule),
    counterfactuals: [],
  };

  return updateImportance(trace);
}

File: optimizer/src/decision_analysis/extract_decisions.py
from __future__ import annotations

import json
from collections import defaultdict
from datetime import datetime
from pathlib import Path

import pandas as pd

from optimization.generate_candidates import CONFIG, EXTERNAL_DIR, OPTIMIZER_ROOT, load_external_variables
from optimization.simulate_field import FieldSimulator


TRACE_PATH = OPTIMIZER_ROOT / "outputs" / "decision_trace.json"


def _json_safe(value):
    if isinstance(value, pd.Timestamp):
        return value.isoformat()
    if isinstance(value, dict):
        return {k: _json_safe(v) for k, v in value.items()}
    if isinstance(value, list):
        return [_json_safe(v) for v in value]
    if hasattr(value, "item"):
        return value.item()
    return value


def _state_diff(before, after):
    changes = {}

    for key in sorted(set(before) | set(after)):
        a = before.get(key)
        b = after.get(key)

        if isinstance(a, (int, float)) and isinstance(b, (int, float)):
            delta = round(float(b) - float(a), 4)
            if abs(delta) > 1e-9:
                changes[key] = {
                    "before": round(float(a), 4),
                    "after": round(float(b), 4),
                    "change": delta,
                }
        elif a != b:
            changes[key] = {
                "before": a,
                "after": b,
            }

    return changes


def simulate_selected_option(data, option):
    field_id = str(option["field_id"])
    management = data["management"].copy()
    management["plan_id"] = management["plan_id"].astype(str)
    plan_map = {str(row["plan_id"]): row for _, row in management.iterrows()}

    simulator = FieldSimulator(data, CONFIG, field_id)
    records = []

    for raw_action in sorted(option["actions"], key=lambda x: pd.Timestamp(x["start_time"])):
        action = dict(raw_action)
        plan_id = str(action["plan_id"])
        start = pd.Timestamp(action["start_time"])

        simulator.advance_to(start)
        before = simulator.snapshot()
        result = simulator.evaluate_and_apply(action, plan_map[plan_id])

        if result is None:
            continue

        after = simulator.snapshot()

        records.append({
            "plan_id": plan_id,
            "candidate_id": str(action["candidate_id"]),
            "field_id": field_id,
            "operation": str(action["operation"]),
            "target": action.get("target", ""),
            "start_time": start,
            "end_time": pd.Timestamp(action["end_time"]),
            "state_before": before,
            "state_after": after,
            "state_changes": _state_diff(before, after),
            "direct_revenue_aud": float(result["direct_revenue_aud"]),
            "direct_cost_aud": float(result["direct_cost_aud"]),
            "direct_cash_effect_aud": float(result["direct_cash_effect_aud"]),
            "state_yield_effect_t_ha": float(result["state_yield_effect_t_ha"]),
        })

    return records


def build_action_records(data, candidates, options, schedule, summary):
    management = data["management"].copy()
    management["plan_id"] = management["plan_id"].astype(str)

    schedule = schedule.copy()
    if not schedule.empty:
        schedule["plan_id"] = schedule["plan_id"].astype(str)
        schedule["candidate_id"] = schedule["candidate_id"].astype(str)

    selected_option_ids = {
        str(item["option_id"])
        for item in summary["selected_options"]
    }

    selected_options = {
        str(option["option_id"]): option
        for option in options
        if str(option["option_id"]) in selected_option_ids
    }

    transitions = {}
    for option in selected_options.values():
        for record in simulate_selected_option(data, option):
            transitions[record["plan_id"]] = record

    records = []

    for _, plan in management.iterrows():
        plan_id = str(plan["plan_id"])
        selected_rows = schedule[schedule["plan_id"] == plan_id] if not schedule.empty else pd.DataFrame()
        selected = not selected_rows.empty
        selected_row = selected_rows.iloc[0] if selected else None

        plan_candidates = candidates[candidates["plan_id"].astype(str) == plan_id].copy()
        if not plan_candidates.empty:
            plan_candidates["start_time"] = pd.to_datetime(plan_candidates["start_time"])
            plan_candidates = plan_candidates.sort_values("start_time")

        timing = {
            "feasible_candidate_count": int(len(plan_candidates)),
            "earliest_feasible_time": None,
            "latest_feasible_time": None,
            "selected_time": None,
            "selected_candidate_id": None,
            "alternative_times": [],
        }

        if not plan_candidates.empty:
            timing["earliest_feasible_time"] = plan_candidates.iloc[0]["start_time"]
            timing["latest_feasible_time"] = plan_candidates.iloc[-1]["start_time"]

        if selected:
            timing["selected_time"] = pd.Timestamp(selected_row["start_time"])
            timing["selected_candidate_id"] = str(selected_row["candidate_id"])

            alternatives = plan_candidates[
                plan_candidates["candidate_id"].astype(str) != str(selected_row["candidate_id"])
            ]

            if not alternatives.empty:
                alternatives = alternatives.assign(
                    distance=(
                        alternatives["start_time"] - pd.Timestamp(selected_row["start_time"])
                    ).abs()
                ).sort_values("distance").head(5)

                timing["alternative_times"] = [
                    {
                        "candidate_id": str(row["candidate_id"]),
                        "start_time": row["start_time"],
                        "direct_cash_effect_aud": float(row["direct_cash_effect_aud"]),
                    }
                    for _, row in alternatives.iterrows()
                ]

        record = {
            "decision_id": f"{plan_id}:action",
            "type": "ACTION_SELECTED" if selected else "ACTION_SKIPPED",
            "plan_id": plan_id,
            "field_id": str(plan["field_id"]),
            "operation": str(plan["operation"]),
            "target": "" if pd.isna(plan["target"]) else str(plan["target"]),
            "required": bool(plan["required"]),
            "selected": selected,
            "depends_on": "" if pd.isna(plan["depends_on"]) else str(plan["depends_on"]),
            "min_gap_hours": 0.0 if pd.isna(plan["min_gap_hours"]) else float(plan["min_gap_hours"]),
            "timing": timing,
            "financial": None,
            "state_transition": transitions.get(plan_id),
            "counterfactual": None,
            "importance": 0.0,
        }

        if selected:
            record["financial"] = {
                "direct_revenue_aud": float(selected_row["direct_revenue_aud"]),
                "direct_cost_aud": float(selected_row["direct_cost_aud"]),
                "direct_cash_effect_aud": float(selected_row["direct_cash_effect_aud"]),
                "state_yield_effect_t_ha": float(selected_row["state_yield_effect_t_ha"]),
            }

        records.append(record)

    return records


def build_field_records(options, summary):
    selected_ids = {
        str(item["option_id"]): item
        for item in summary["selected_options"]
    }

    options_by_field = defaultdict(list)
    for option in options:
        options_by_field[str(option["field_id"])].append(option)

    records = []

    for field_id, field_options in options_by_field.items():
        selected = next(
            option for option in field_options
            if str(option["option_id"]) in selected_ids
        )

        alternatives = [
            option for option in field_options
            if str(option["option_id"]) != str(selected["option_id"])
        ]
        alternatives.sort(
            key=lambda x: float(x["objective_value_aud"]),
            reverse=True,
        )

        best_alt = alternatives[0] if alternatives else None

        records.append({
            "decision_id": f"{field_id}:field_option",
            "type": "FIELD_OPTION_SELECTED",
            "field_id": field_id,
            "selected_option_id": str(selected["option_id"]),
            "selected_objective_value_aud": float(selected["objective_value_aud"]),
            "selected_direct_cash_effect_aud": float(selected["total_direct_cash_effect_aud"]),
            "selected_terminal_value_aud": float(selected["terminal_value_aud"]),
            "best_local_alternative": None if best_alt is None else {
                "option_id": str(best_alt["option_id"]),
                "objective_value_aud": float(best_alt["objective_value_aud"]),
                "local_value_difference_aud": round(
                    float(selected["objective_value_aud"])
                    - float(best_alt["objective_value_aud"]),
                    2,
                ),
            },
        })

    return records


def build_resource_records(data, schedule):
    if schedule.empty:
        return []

    schedule = schedule.copy()
    schedule["start_time"] = pd.to_datetime(schedule["start_time"])
    schedule["end_time"] = pd.to_datetime(schedule["end_time"])
    records = []

    for machine_id, group in schedule.groupby("machine_id"):
        used_hours = sum(
            (row["end_time"] - row["start_time"]).total_seconds() / 3600.0
            for _, row in group.iterrows()
        )

        machine_availability = data["machine_availability"][
            (data["machine_availability"]["machine_id"] == machine_id)
            & (data["machine_availability"]["available"] == 1)
        ]

        available_hours = 0.0
        for _, row in machine_availability.iterrows():
            start = pd.Timestamp(f"{row['date']} {row['available_from']}")
            end = pd.Timestamp(f"{row['date']} {row['available_to']}")
            available_hours += max(0.0, (end - start).total_seconds() / 3600.0)

        utilisation = used_hours / available_hours if available_hours > 0 else 0.0

        records.append({
            "decision_id": f"machine:{machine_id}",
            "type": "RESOURCE_UTILISATION",
            "resource_type": "machine",
            "resource_id": str(machine_id),
            "used_hours": round(used_hours, 2),
            "available_hours": round(available_hours, 2),
            "utilisation": round(utilisation, 4),
            "binding": utilisation >= 0.90,
            "affected_plans": group["plan_id"].astype(str).tolist(),
        })

    for date, row in data["water"].set_index("date").iterrows():
        used = schedule.loc[
            schedule["start_time"].dt.date == date,
            "water_ml",
        ].sum()

        capacity = min(
            float(row["available_water_ml"]),
            float(row["max_delivery_ml_per_day"]),
        )

        utilisation = float(used) / capacity if capacity > 0 else 0.0

        records.append({
            "decision_id": f"water:{date}",
            "type": "RESOURCE_UTILISATION",
            "resource_type": "water",
            "date": str(date),
            "used_ml": round(float(used), 3),
            "capacity_ml": round(capacity, 3),
            "utilisation": round(utilisation, 4),
            "binding": utilisation >= 0.90,
        })

    time_step = pd.Timedelta(hours=float(CONFIG["candidate_generation"]["time_step_hours"]))
    start = schedule["start_time"].min().floor("h")
    end = schedule["end_time"].max().ceil("h")
    t = start

    while t < end:
        labour_rows = data["labour"][data["labour"]["date"] == t.date()]

        if not labour_rows.empty:
            capacity = int(labour_rows.iloc[0]["available_workers"])
            active = schedule[
                (schedule["start_time"] <= t)
                & (schedule["end_time"] > t)
            ]

            used = int(active["workers_required"].sum())
            utilisation = used / capacity if capacity > 0 else 0.0

            if used > 0 or utilisation >= 0.90:
                records.append({
                    "decision_id": f"labour:{t.isoformat()}",
                    "type": "RESOURCE_UTILISATION",
                    "resource_type": "labour",
                    "time": t,
                    "used_workers": used,
                    "available_workers": capacity,
                    "utilisation": round(utilisation, 4),
                    "binding": utilisation >= 0.90,
                    "affected_plans": active["plan_id"].astype(str).tolist(),
                })

        t += time_step

    return records


def update_importance(trace):
    baseline = abs(float(trace["summary"]["total_objective_value_aud"])) or 1.0

    counterfactual_by_plan = {
        str(x["plan_id"]): x
        for x in trace.get("counterfactuals", [])
    }

    for record in trace["actions"]:
        score = 0.0

        if record["required"]:
            score += 0.15

        financial = record.get("financial")

        if financial and financial["direct_cash_effect_aud"] < 0:
            score += min(
                0.20,
                abs(float(financial["direct_cash_effect_aud"])) / baseline * 10,
            )

        counterfactual = counterfactual_by_plan.get(record["plan_id"])

        if counterfactual:
            record["counterfactual"] = counterfactual

            objective_change = counterfactual.get("objective_change_aud")

            if objective_change is not None:
                delta = abs(float(objective_change))
                score += min(0.60, delta / baseline * 20)

            elif not counterfactual.get("feasible", True):
                # If forcing/forbidding the action makes the whole schedule
                # infeasible, that is itself an important decision.
                score += 0.60

        if not record["required"]:
            score += 0.10

        record["importance"] = round(min(1.0, score), 4)

    return trace


def generate_decision_trace(
    candidates,
    options,
    schedule,
    summary,
    external_variables_dir=EXTERNAL_DIR,
):
    data = load_external_variables(external_variables_dir)

    trace = {
        "generated_at": datetime.now().isoformat(),
        "summary": summary,
        "actions": build_action_records(
            data,
            candidates,
            options,
            schedule,
            summary,
        ),
        "field_decisions": build_field_records(options, summary),
        "resources": build_resource_records(data, schedule),
        "counterfactuals": [],
    }

    return update_importance(trace)


def save_decision_trace(trace, path=TRACE_PATH):
    path = Path(path)
    path.parent.mkdir(parents=True, exist_ok=True)

    with open(path, "w", encoding="utf-8") as f:
        json.dump(_json_safe(trace), f, indent=2)
